src/Storages/PostgreSQL/PostgreSQLConnectionPool.cpp

```diff
@@ -0,0 +1,67 @@
+#if !defined(ARCADIA_BUILD)
+#include "config_core.h"
+#endif
+
+#if USE_LIBPQXX
+#include <IO/WriteBufferFromString.h>
+#include <IO/Operators.h>
+#include "PostgreSQLConnectionPool.h"
+#include "PostgreSQLConnection.h"
+#include <common/logger_useful.h>
+
+
+namespace DB
+{
+
+PostgreSQLConnectionPool::PostgreSQLConnectionPool(
+    std::string dbname, std::string host, UInt16 port, std::string user, std::string password)
+    : pool(std::make_shared<Pool>(POSTGRESQL_POOL_DEFAULT_SIZE))
+{
+    address = host + ':' + std::to_string(port);
+    connection_str = formatConnectionString(std::move(dbname), std::move(host), port, std::move(user), std::move(password));
+}
+
+
+PostgreSQLConnectionPool::PostgreSQLConnectionPool(const PostgreSQLConnectionPool & other)
+        : pool(std::make_shared<Pool>(POSTGRESQL_POOL_DEFAULT_SIZE))
+        , connection_str(other.connection_str)
+        , address(other.address)
+{
+}
+
+
+std::string PostgreSQLConnectionPool::formatConnectionString(
+    std::string dbname, std::string host, UInt16 port, std::string user, std::string password)
+{
+    WriteBufferFromOwnString out;
+    out << "dbname=" << quote << dbname
+        << " host=" << quote << host
+        << " port=" << port
+        << " user=" << quote << user
+        << " password=" << quote << password;
+    return out.str();
+}
+
+
+PostgreSQLConnectionHolderPtr PostgreSQLConnectionPool::get()
+{
+    PostgreSQLConnectionPtr connection;
+    if (pool->tryPop(connection, POSTGRESQL_POOL_WAIT_MS))
+    {
+        return std::make_shared<PostgreSQLConnectionHolder>(connection, *pool);
+    }
+
+    connection = std::make_shared<PostgreSQLConnection>(connection_str, address);
```

Comment:
This basically means that the pool grows without limit under a load spike. After it has grown to a couple of hundred connections, in some way it becomes worse that no poll -- the number of simultaneous connections has a high limit, and we never close these connections. We still save the latency on establishing the connection, though.

Maybe it's better to have a long wait here, and error out if there is still no free connection?

Reply:
The pool does not grow, because ConcurrentBoundedQueue has fixed size and in PostgreSQLConnectionHolder I use `tryPush` to queue, which will put connection back only if there is some place, otherwise it is closed.

Reply:
You're right, I missed that. Maybe some comment describing this will be helpful.

Postgres doesn't like high number of concurrent connections, so people might want to use this pool to limit this number, but it won't work, right? I'm not sure whether this is an important case and we should support it.